Pytorch cotainr (#4)
DeiC-HPC/BitBulldozersLab: pytorch-cotainr/README.md

> # Custom PyTorch with additional libraries

Reviewer: Date and keyword specifiers are missing (see the [BitBulldozersLab README file](https://github.com/DeiC-HPC/BitBulldozersLab/blob/main/README.md#poc--example-readmemd) for the requirements for the content of this README)

> # Custom PyTorch with additional libraries

Reviewer: I think this is more accurate title for what this PoC is about:
```suggestion
# Building the LUMI PyTorch container using cotainr
```

The PyTorch container on LUMI are custom built with a variety of custom wheels and source code from

> speciala ROCm repositories. An example can be seen on [GitHub](https://github.com/sfantao/lumi-containers/blob/lumi-sep2024/pytorch/build-rocm-6.0.3-python-3.12-pytorch-v2.3.1.docker).

Reviewer: Given all the trouble in building these containers, I think we should fix the scope of this BitBulldozer task to just trying to build one specific Pytorch container, e.g. the v2.3.1 based on ROCm 6.0.3 (or altenatively, v2.4.1 based on ROCm 6.1.3 given your comments below about needing a more recent ROCm version for bitsandbytes). You have essentially already done this, so this is mostly a comment about documenting it as such.
```suggestion
The PyTorch v2.3.1 container based on ROCm 6.0.3 on LUMI (i.e. `/appl/local/containers/tested-containers/lumi-pytorch-rocm-6.0.3-python-3.12-pytorch-v2.3.1-dockerhash-2c1c14cafd28.sif`) is custom built with a variety of custom wheels and source code from special ROCm repositories as defined in the Dockerfile in https://github.com/sfantao/lumi-containers/blob/lumi-sep2024/pytorch/build-rocm-6.0.3-python-3.12-pytorch-v2.3.1.docker.
```

> This repo attempts to build a similar container using a conda_env.yml file and cotainr, the test-container is built using the command in `build.txt`.

Reviewer: The `conda_env.yml` file is not included in the repo. I guess it's supposed to be an attempt at including everything in Samuel's container in a single conda environment file? (There are also multiple mentions of this `conda_env.yml` file later on in this README).

Also, it seems that one needs to run the `export_env.sh` script before building at least some of the containers. I think it might be better to just in-line the necessary commands (in this README) needed to build the container for each of the different test-containers and avoid the `build.txt` file altogether.

Author: I have reformated to no longer refer to `conda_env.yml` (which originally referred to the relevant `conda_env_*.yml`)

> ## Running this BitBulldozer

Reviewer: Maybe rename this section to "Building the containers" or similar. And then have subsections for:
- setting up the prerequisites
- building the containers (possibly one subsection for each)

Author: I have added this.

> We first `git clone https://github.com/DeiC-HPC/BitBulldozersLab.git` this repo to LUMI.

Reviewer: Making this very tutorial-like is not strictly required. You could make this description more compact, if you like. Just FYI.

Author: FYI I find this advice somewhat to contradicts the other advice on "Maybe rename this section to "Building the containers" or similar", as this change in fact made the README even more tutorial-like imo.

> This repo used the 2023.11.0 cotainr release, which is included in this repo as a submodule. It is an empty folder by default, and requires initialization 

Reviewer: Maybe make it a little more explicit that the builds are done on LUMI and everything here is happening on LUMI.

Author: I thought the previous sentence "We first `git clone https://github.com/DeiC-HPC/BitBulldozersLab.git` this repo to LUMI." was pretty explicit.

```
$ git submodule init
$ git submodule update
```
Before we are ready to run `cotainr` we need to have a sufficiently recent Python executable, so we run
```
module load cray-python
```
Then, we are ready to test various packages, the neccesary commands are included in `build.txt` which copy-pasted and run. Note, that in order to get the unfiltered error message from `cotainr`, we have to change the following line

> ```diff

Reviewer: Is this really necessary? Or can you get away with running cotainr with `-vvv` for [singularity debug output](https://cotainr.readthedocs.io/en/stable/development/tracing_logging.html#cotainr-tracing-log-levels)? It also makes the output from conda a little more noisy, but it seems to provide the necessary error messages as part of the output when a failure happens.

If `-vvv` is sufficient, I think we should avoid using a git submodule and just load the relevant cotainr module on LUMI.

Author: The `git submodule` is not used for the purpose of applying this patch, but rather to lock down the specific cotainr version (commit) to ensure reproducibility in a self-contained way.

cotainr/cotainr/container.py#L258
- [line for line in e.stderr.split("\n") if line.startswith("FATAL")]
+ [line for line in e.stderr.split("\n")]

```
The success library most closely resembling the [Container by Samuel](https://github.com/sfantao/lumi-containers/blob/lumi-sep2024/pytorch/build-rocm-6.1.3-python-3.12-pytorch-v2.4.1.docker), is run via
```
cotainr/bin/cotainr build --base-image /appl/local/containers/sif-images/lumi-rocm-rocm-6.1.3.sif --conda-env conda_env_success.yml test.sif -vv --accept-licenses
```
Some basic import tests of key libraries result in the following warnings:
```
Singularity> python
Python 3.12.8 | packaged by conda-forge | (main, Dec  5 2024, 14:24:40) [GCC 13.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> import deepspeed

> [2025-02-05 11:40:34,904] [WARNING] [real_accelerator.py:162:get_accelerator] Setting accelerator to CPU. If you have GPU or other accelerator, we were unable to detect it.

Reviewer: Was this run on a login node? I think, it should still pick up a GPU, if run on a LUMI-G node, despite the C-extensions for deepspeed not being pre-compiled in the container.

Author: It was run on a login node, I have updated with the output on LUMI-G

[2025-02-05 11:40:34,965] [INFO] [real_accelerator.py:203:get_accelerator] Setting ds_accelerator to cpu (auto detect)
>>> import bitsandbytes

> /opt/conda/envs/conda_container_env/lib/python3.12/site-packages/bitsandbytes/backends/cpu_xpu_common.py:29: UserWarning: g++ not found, torch.compile disabled for CPU/XPU.

Reviewer: I guess this warning could be handled by having a base image that defines `g++` or by conda installing a compiler package that provides `g++`. However, it probably doesn't make much difference since it seems to only result in the torch JIT compiler being disabled, which is probably not that big a deal since the JIT compiler probably attempts to add files to the immutable container if used. It may still be worth mentioning, though.

Author: I have added a comment

  warnings.warn("g++ not found, torch.compile disabled for CPU/XPU.")
```

> ## Library removal concerns

Reviewer: Maybe have a section named "Problems & concerns" (or similar) with a short paragraph stating that these are the problems and concerns encountered when trying to build the containers. And then have "Library removal concerns", "Writing pip commands in conda-env", "Conda yaml deficiencies" as sub-sections to that "Problems & concerns" section.

Author: I have added this

> After installation of all the packages, the _Conda_ libstc++.so library  is explicitly [removed](https://github.com/sfantao/lumi-containers/blob/lumi-sep2024/common/Dockerfile.no-torch-libstdc%2B%2B) from the container to ensure the container C++ library is used. This might or might not be an issue for portability of the container depending on how the base container is built. It nevertheless is an issue for the cotainer method as it doesn't support modifying the container post-install. This is done in all versions of the docker script. Note: There are similar concerns for ROCm>=6.1.3 versions (without LLVM), where the ROCm libraries from the PyTorch installation are removed.

Reviewer: ```suggestion
After installation of all the packages, the _Conda_ libstc++.so library  is explicitly [removed](https://github.com/sfantao/lumi-containers/blob/lumi-sep2024/common/Dockerfile.no-torch-libstdc%2B%2B) from the official LUMI PyTorch container to ensure the container C++ library is used. This might or might not be an issue for portability of the container depending on how the base container is built. It nevertheless is an issue for the cotainr method as it doesn't support modifying the container post-install. This is done in all versions of the docker script. Note: There are similar concerns for ROCm>=6.1.3 versions (without LLVM), where the ROCm libraries from the PyTorch installation are removed.
```

## Conda yaml deficiencies
The attempted cotainr build with `conda_env.yml` in this repo failed, whereas the cotainr build with `conda_env_ref.yml` and subsequent manual installing the `pip install apex @ git+https://github.com/rocm/apex` does succeed. The reason for this failure is described [here](https://github.com/astral-sh/uv/issues/1715) as:

_A number of machine learning-related modules Flash-attn, NVidia Apex need pytorch to be installed before they are installed, so that they can do things like compile against version of pytorch currently installed in the user's environment._

> _You might think that these modules should declare that they depend upon Pytorch at setup time in their pyproject.toml file, but that isn't a good solution. No one wants a newly installed Pytorch version, and they might have installed pytorch using Conda or some other mechanism that pip isn't aware of. And these modules want to just work with whatever version of pytorch is already there. It would cause a LOT more problems than it would solve._

Reviewer: In general, I don't agree with this statement. If you tap into the Python build ecosystem, e.g. via setup.py and/or pyproject.toml, I think you also need to implement the relevant PEPs, e.g. [PEP528](https://peps.python.org/pep-0518/) about properly specifying all build dependencies. I think its reasonable to expect that a package that is pip installable, is actually able to install correctly in some way without the users having to care about multi-step install procedures to manually install dependencies in the correct order. I understand the that there are use cases where this causes problem, but then you have options like `--no-build-isolation` to opt out and handle the dependencies installation manually.

Anyway, the important message here is that in order to use a tool like cotainr, you need to be able to declaratively specify all needed packages and they must be installable in a one-step procedure. If some packages don't explicitly declare build dependencies, this is not achievable, which is the main problem here, I think.

Author: It is a direct quote. I am impartial on whether or not I agree with the statement. However, to me, the statement simply establishes that it seems to be a conscious decision from the ML community to diverge from the Python build ecosystem. I can remove it if you for some reason find it inappropriate to directly quote.

Reviewer: I think from our point of view (at least mine), the most important findings here are:
- Cotainr by design relies on package managers to be able to resolve dependencies correctly.
- For **build time** dependency management to work in the Python package ecosystem, packages must implement PEP528. Only declaring install time dependencies is insufficient to ensure correct dependency handling if pip (or another tool) tries to build a wheel from source at install time.
- It is common for PyTorch "extension packages" (e.g. Apex, Deepspeed, bitandbytes) to not explicitly declare PyTorch as a build time dependencies according to PEP528 but instead rely on the user to manually install PyTorch before attempting to install these packages. Installing such packages from a pip requirement.txt file fails. 

Consequently, cotainr also fails to install these packages when added as pip requirements in a conda environment file making it impossible to directly build Samuel's PyTorch container using cotainr.

I would like the main emphasis on the discussion to be about this problem when using cotainr. Then we can also briefly discuss that there are reasons why packages avoid PEP528, but I would maybe not include the quote and only link the comment as a reference for reasons to avoid PEP528.

Author: Thanks, I have added the changes. I like your summary, so I have added it to the README. I hope you don't mind!

> Essentially, the package `Apex` **needs** to be built with `--no-build-isolation` after PyTorch is installed in a two-step process. However, because conda installs all requested dependencies at once using a generated `requirements.txt` file it would be impossible without additional steps. This could possibly be accomplished by allowing separately pip installing a `requirements.txt` after the conda installation like suggested in this [PR](https://github.com/DeiC-HPC/cotainr/pull/55). We would need to be careful that the pip installation properly uses the conda environment. Note, that this issue is not unique to the ROCm branch, but is required with Nvidia as well.

Reviewer: Possible alternatives (not really solutions, but workarounds), that we might want to just list for future reference, are:
- Using some other tool (e.g. uv) for installing the python environment, if such a tool allows us to specify a specific  order of installation or otherwise overcome this problem
- Installing the "problematic" dependency (torch) using conda, e.g. by [converting the pip wheel to a conda package](https://docs.conda.io/projects/conda-build/en/stable/user-guide/wheel-files.html). This would still require a pre-processing step of doing the conversion, though.
- Convince AMD and/or the individual package maintainers to build binaries (e.g. pip wheels or conda packages) with consistent specification of runtime dependencies for these packages. (I believe Nvidia provides such packages via https://anaconda.org/nvidia/repo for those packages that are not already on conda-forge.)
